### PR TITLE
[Bugfix][Frontend] Fix missing `/metrics` endpoint

### LINKED ISSUE #6461: [Bug]: No metrics exposed at /metrics with 0.5.2 (0.5.1 is fine), possible regression?
### Your current environment

```text
Collecting environment information...                                                     
PyTorch version: 2.3.1+cu121                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
Is debug build: False                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
CUDA used to build PyTorch: 12.1                                                          
ROCM used to build PyTorch: N/A                                                                                                                                                      
                                                                                                                                                                                     
OS: Ubuntu 22.04.4 LTS (x86_64)                                                                                                                                                      
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0                                        
Clang version: Could not collect                                                          
CMake version: version 3.30.0                                                             
Libc version: glibc-2.35                                                                  
                                                                                                                                                                                     
Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)                                                                                                  
Python platform: Linux-5.15.0-107-generic-x86_64-with-glibc2.35                                                                                                                      
Is CUDA available: True                                                                   
CUDA runtime version: Could not collect                                                   
CUDA_MODULE_LOADING set to: LAZY                                                          
GPU models and configuration:                                                             
GPU 0: NVIDIA H100 80GB HBM3                                                              
  MIG 3g.40gb     Device  0:                                                              
                                                                                          
Nvidia driver version: 550.54.15                                                          
cuDNN version: Could not collect                                                          
HIP runtime version: N/A                                                                  
MIOpen runtime version: N/A                                                               
Is XNNPACK available: True                                                                
                                                                                          
CPU:                                                                                      
Architecture:                       x86_64                                                
CPU op-mode(s):                     32-bit, 64-bit                                        
Address sizes:                      46 bits physical, 57 bits virtual                     
Byte Order:                         Little Endian                                                                                                                                    
CPU(s):                             192                                                                                                                                              
On-line CPU(s) list:                0-191                                                                                                                                            
Vendor ID:                          GenuineIntel                                                                                                                                     
Model name:                         Intel(R) Xeon(R) Platinum 8468                        
CPU family:                         6                                                     
Model:                              143                                                   
Thread(s) per core:                 2                                                                                                                                                                                                                        
Core(s) per socket:                 48                                                                                                                                                                                                                       
Socket(s):                          2                                                                                                                                                                                                                        
Stepping:                           8                                                                                                                                                                                                                        
CPU max MHz:                        3800.0000                                                                                 
CPU min MHz:                        800.0000                                                                                  
BogoMIPS:                           4200.00                                                                                   
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat
_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke
 waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                     VT-x                                                                                      
L1d cache:                          4.5 MiB (96 instances)                                                                    
L1i cache:                          3 MiB (96 instances)                                                                      
L2 cache:                           192 MiB (96 instances)                                                                    
L3 cache:                           210 MiB (2 instances)                                                                     
NUMA node(s):                       2                                                                                         
NUMA node0 CPU(s):                  0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94,96,98,100,102,104,106,108,110,112,114,116,118,120,122,124,126,128,130,132,134,136,138,140,142,144,146,148,150,152,154,156,158,160,162,164,166,168,170,172,174,176,178,180,182,184,186,188,190
NUMA node1 CPU(s):                  1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95,97,99,101,103,105,107,109,111,113,115,117,119,121,123,125,127,129,131,133,135,137,139,141,143,145,147,149,151,153,155,157,159,161,163,165,167,169,171,173,175,177,179,181,183,185,187,189,191
Vulnerability Gather data sampling: Not affected                                                                              
Vulnerability Itlb multihit:        Not affected                                                                              
Vulnerability L1tf:                 Not affected                                                                              
Vulnerability Mds:                  Not affected                                                                              
Vulnerability Meltdown:             Not affected                                                                              
Vulnerability Mmio stale data:      Not affected                                                                              
Vulnerability Retbleed:             Not affected                                                                              
Vulnerability Spec rstack overflow: Not affected                                                                              
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp                                                                                                                                                      
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization                                                                                                                                                     
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S                                                                                                                         
Vulnerability Srbds:                Not affected                                                                              
Vulnerability Tsx async abort:      Not affected                                                                              

Versions of relevant libraries:                                                                                               
[pip3] flashinfer==0.0.9+cu121torch2.3                                                                                        
[pip3] numpy==1.26.4                                                                                                          
[pip3] nvidia-nccl-cu12==2.20.5                                                                                               
[pip3] torch==2.3.1                                                                                                           
[pip3] torchvision==0.18.1                                                                                                    
[pip3] transformers==4.42.4                                                                                                   
[pip3] triton==2.3.1                                                                                                          
[conda] Could not collect                                                                                                     
ROCM Version: Could not collect                                                                                               
Neuron SDK Version: N/A                                                                                                       
vLLM Version: 0.5.2                                                                                                           
vLLM Build Flags:                                                                                                             
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled                                                                         
GPU Topology:                                                                                                                 
GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID                                                                           
GPU0     X      1,3,5,7,9,11    1               N/A                                                                           

Legend:                                                                                                                       

  X    = Self                                                                                                                 
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)                                                                                                                                                       
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node                                                                                                                                                 
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)                                                                                                                                                                        
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)                                                                                                                                                               
  PIX  = Connection traversing at most a single PCIe bridge                                                                   
  NV#  = Connection traversing a bonded set of # NVLinks
```


### 🐛 Describe the bug

When switching the Container image from `0.5.1` to `0.5.2` the Prometheus metrics usually exposed at /metrics are gone.

0.5.1 - `INFO:     xxx.xxx.xxx.xxx:56742 - "GET /metrics HTTP/1.1" 200 OK`
0.5.2 - `INFO:     xxx.xxx.xxx.xxx:45698 - "GET /metrics HTTP/1.1" 404 Not Found`



I looked at the release notes, but did not see any change about a required config parameter or similar to have the metrics enabled.

### PR DESCRIPTION
**Pull request by DarkLight1337**
Mount does not work on router, so I'm reverting the change in #5090 so that mount is performed on the application object. Since `app` is no longer global, I've moved the code into a function to be called after `app` is constructed.

To avoid similar issues in the future, I've added some tests to check that the endpoints remain accessible.

FIX #6461

### CODE REVIEW

**Comment by DarkLight1337** [comment]
@robertgshaw2-neuralmagic can you help update the test to check #4523? Since I haven't worked with prometheus metrics before, I'm worried that I might break something with this refactor.

File: tests/entrypoints/openai/test_basic.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        "bfloat16",
+        "--max-model-len",
+        "8192",
+        "--enforce-eager",
+        "--max-num-seqs",
+        "128",
+    ]
+
+    with RemoteOpenAIServer(args) as remote_server:
+        yield remote_server
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply tests/entrypoints/openai/test_basic.py:30]: If #6431 is merged before this one, please apply these changes.

**Review by simon-mo — APPROVED** [review]

**Comment by DarkLight1337** [comment]
@robertgshaw2-neuralmagic since this PR has been merged already, you can open a new PR if you need to update the tests to be more robust.

**Comment by xfalcox** [comment]
Can we get a new docker release with this fix?

**Comment by servient-ashwin** [comment]
@DarkLight1337  
This fix is available as per the `0.5.3`  [log](https://github.com/vllm-project/vllm/releases/tag/v0.5.3) but the available endpoints doesn't show it
```
INFO 07-25 20:37:21 api_server.py:292] Available routes are:
INFO 07-25 20:37:21 api_server.py:297] Route: /openapi.json, Methods: GET, HEAD
INFO 07-25 20:37:21 api_server.py:297] Route: /docs, Methods: GET, HEAD
INFO 07-25 20:37:21 api_server.py:297] Route: /docs/oauth2-redirect, Methods: GET, HEAD
INFO 07-25 20:37:21 api_server.py:297] Route: /redoc, Methods: GET, HEAD
INFO 07-25 20:37:21 api_server.py:297] Route: /health, Methods: GET
INFO 07-25 20:37:21 api_server.py:297] Route: /tokenize, Methods: POST
INFO 07-25 20:37:21 api_server.py:297] Route: /detokenize, Methods: POST
INFO 07-25 20:37:21 api_server.py:297] Route: /v1/models, Methods: GET
INFO 07-25 20:37:21 api_server.py:297] Route: /version, Methods: GET
INFO 07-25 20:37:21 api_server.py:297] Route: /v1/chat/completions, Methods: POST
INFO 07-25 20:37:21 api_server.py:297] Route: /v1/completions, Methods: POST
INFO 07-25 20:37:21 api_server.py:297] Route: /v1/embeddings, Methods: POST
```

was it renamed to health?

**Comment by ZeroYuJie** [comment]
v0.5.4 docker release still missing /metrics endpoint

**Comment by ZeroYuJie** [comment]
I try use main branch to build docker image,The metrics information has become completely unavailable,no longer contains any statistical information related to the model.
```
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 7313.0
python_gc_objects_collected_total{generation="1"} 5988.0
python_gc_objects_collected_total{generation="2"} 45.0
# HELP python_gc_objects_uncollectable_total Uncollectable objects found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 922.0
python_gc_collections_total{generation="1"} 83.0
python_gc_collections_total{generation="2"} 6.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="10",patchlevel="14",version="3.10.14"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.3022793728e+010
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 7.00747776e+08
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.72291887821e+09
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 18.240000000000002
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 22.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1.048576e+06
```

**Comment by DarkLight1337** [comment]
@robertgshaw2-neuralmagic is this due to the 

> I try use main branch to build docker image,The metrics information has become completely unavailable,no longer contains any statistical information related to the model.
> 
> ```
> # HELP python_gc_objects_collected_total Objects collected during gc
> # TYPE python_gc_objects_collected_total counter
> python_gc_objects_collected_total{generation="0"} 7313.0
> python_gc_objects_collected_total{generation="1"} 5988.0
> python_gc_objects_collected_total{generation="2"} 45.0
> # HELP python_gc_objects_uncollectable_total Uncollectable objects found during GC
> # TYPE python_gc_objects_uncollectable_total counter
> python_gc_objects_uncollectable_total{generation="0"} 0.0
> python_gc_objects_uncollectable_total{generation="1"} 0.0
> python_gc_objects_uncollectable_total{generation="2"} 0.0
> # HELP python_gc_collections_total Number of times this generation was collected
> # TYPE python_gc_collections_total counter
> python_gc_collections_total{generation="0"} 922.0
> python_gc_collections_total{generation="1"} 83.0
> python_gc_collections_total{generation="2"} 6.0
> # HELP python_info Python platform information
> # TYPE python_info gauge
> python_info{implementation="CPython",major="3",minor="10",patchlevel="14",version="3.10.14"} 1.0
> # HELP process_virtual_memory_bytes Virtual memory size in bytes.
> # TYPE process_virtual_memory_bytes gauge
> process_virtual_memory_bytes 1.3022793728e+010
> # HELP process_resident_memory_bytes Resident memory size in bytes.
> # TYPE process_resident_memory_bytes gauge
> process_resident_memory_bytes 7.00747776e+08
> # HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
> # TYPE process_start_time_seconds gauge
> process_start_time_seconds 1.72291887821e+09
> # HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
> # TYPE process_cpu_seconds_total counter
> process_cpu_seconds_total 18.240000000000002
> # HELP process_open_fds Number of open file descriptors.
> # TYPE process_open_fds gauge
> process_open_fds 22.0
> # HELP process_max_fds Maximum number of open file descriptors.
> # TYPE process_max_fds gauge
> process_max_fds 1.048576e+06
> ```

Can you revert to the commit from this PR to ensure that the problem isn't caused by subsequent commits? (Notably #6883)

**Comment by frittentheke** [comment]
The fix https://github.com/vllm-project/vllm/pull/6463/files for the former issue https://github.com/vllm-project/vllm/issues/6461 also contained tests to ensure metrics are there. Are they not being used @DarkLight1337 ?

**Comment by DarkLight1337** [comment]
The test only checks that the metrics endpoint exists. It doesn't check which metrics are being exposed.

**Comment by frittentheke** [comment]
> The test only checks that the metrics endpoint exists. It doesn't check which metrics are being exposed.

That might be a cool thing to test then ;-)

**Pull request merged**